akka-cluster-tools/src/multi-jvm/scala/akka/cluster/pubsub/DistributedPubSubRestartSpec.scala

```diff
@@ -0,0 +1,162 @@
+    akka.cluster.auto-down-unreachable-after = off
+    """))
+
+  testTransport(on = true)
+
+  class Shutdown extends Actor {
+    def receive = {
+      case "shutdown" ⇒ context.system.terminate()
+    }
+  }
+
+}
+
+class DistributedPubSubRestartMultiJvmNode1 extends DistributedPubSubRestartSpec
+class DistributedPubSubRestartMultiJvmNode2 extends DistributedPubSubRestartSpec
+class DistributedPubSubRestartMultiJvmNode3 extends DistributedPubSubRestartSpec
+
+class DistributedPubSubRestartSpec extends MultiNodeSpec(DistributedPubSubRestartSpec) with STMultiNodeSpec with ImplicitSender {
+  import DistributedPubSubRestartSpec._
+  import DistributedPubSubMediator._
+
+  override def initialParticipants = roles.size
+
+  def join(from: RoleName, to: RoleName): Unit = {
+    runOn(from) {
+      Cluster(system) join node(to).address
+      createMediator()
+    }
+    enterBarrier(from.name + "-joined")
+  }
+
+  def createMediator(): ActorRef = DistributedPubSub(system).mediator
+  def mediator: ActorRef = DistributedPubSub(system).mediator
+
+  def awaitCount(expected: Int): Unit = {
+    val probe = TestProbe()
+    awaitAssert {
+      mediator.tell(Count, probe.ref)
+      probe.expectMsgType[Int] should ===(expected)
+    }
+  }
+
+  "A Cluster with DistributedPubSub" must {
+
+    "startup 3 node cluster" in within(15 seconds) {
+      join(first, first)
+      join(second, first)
+      join(third, first)
+      enterBarrier("after-1")
+    }
+
+    "handle restart of nodes with same address" in within(30 seconds) {
+      mediator ! Subscribe("topic1", testActor)
+      expectMsgType[SubscribeAck]
+      awaitCount(3)
+
+      runOn(first) {
+        mediator ! Publish("topic1", "msg1")
+      }
+      enterBarrier("pub-msg1")
+
+      expectMsg("msg1")
+      enterBarrier("got-msg1")
+
+      runOn(second) {
+        mediator ! Internal.DeltaCount
+        val oldDeltaCount = expectMsgType[Long]
+
+        enterBarrier("end")
+
+        mediator ! Internal.DeltaCount
+        val deltaCount = expectMsgType[Long]
+        deltaCount should ===(oldDeltaCount)
+      }
+
+      runOn(first) {
+        mediator ! Internal.DeltaCount
+        val oldDeltaCount = expectMsgType[Long]
+
+        val thirdAddress = node(third).address
+        testConductor.shutdown(third).await
+
+        within(20.seconds) {
+          awaitAssert {
+            system.actorSelection(RootActorPath(thirdAddress) / "user" / "shutdown") ! Identify(None)
+            expectMsgType[ActorIdentity](1.second).ref.get
+          }
+        }
+
+        system.actorSelection(RootActorPath(thirdAddress) / "user" / "shutdown") ! "shutdown"
+
+        enterBarrier("end")
+
+        mediator ! Internal.DeltaCount
+        val deltaCount = expectMsgType[Long]
+        deltaCount should ===(oldDeltaCount)
+      }
+
+      runOn(third) {
+        Await.result(system.whenTerminated, 10.seconds)
+        val newSystem = ActorSystem(
```

Comment:
aren't non-deamon threads preventing the exit? Or is handled by the multi-node framework calling `exit` somewhere?